compiler/noirc_frontend/src/elaborator/unquote.rs

```diff
@@ -1,306 +1,41 @@
-                self.find_unquoted_exprs_in_method_call(call, unquoted_exprs);
-            }
-            ExpressionKind::Constructor(constructor) => {
-                self.find_unquoted_exprs_in_constructor(constructor, unquoted_exprs);
-            }
-            ExpressionKind::MemberAccess(access) => {
-                self.find_unquoted_exprs_in_access(access, unquoted_exprs);
-            }
-            ExpressionKind::Cast(cast) => self.find_unquoted_exprs_in_cast(cast, unquoted_exprs),
-            ExpressionKind::Infix(infix) => {
-                self.find_unquoted_exprs_in_infix(infix, unquoted_exprs);
-            }
-            ExpressionKind::If(if_) => self.find_unquoted_exprs_in_if(if_, unquoted_exprs),
-            ExpressionKind::Variable(_, _) => (),
-            ExpressionKind::Tuple(tuple) => {
-                self.find_unquoted_exprs_in_tuple(tuple, unquoted_exprs);
-            }
-            ExpressionKind::Lambda(lambda) => {
-                self.find_unquoted_exprs_in_lambda(lambda, unquoted_exprs);
-            }
-            ExpressionKind::Parenthesized(expr) => {
-                self.find_unquoted_exprs_in_expr(expr, unquoted_exprs);
-            }
-            ExpressionKind::Quote(quote, _) => {
-                self.find_unquoted_exprs_in_block(quote, unquoted_exprs);
-            }
-            ExpressionKind::Comptime(block, _) => {
-                self.find_unquoted_exprs_in_block(block, unquoted_exprs);
-            }
-            ExpressionKind::Resolved(_) => (),
-            ExpressionKind::Error => (),
-            ExpressionKind::UnquoteMarker(_) => (),
-            ExpressionKind::Unquote(unquoted) => {
-                // Avoid an expensive clone for unquoted
-                let empty_expr = Expression::new(ExpressionKind::Error, unquoted.span);
-                let unquote = std::mem::replace(unquoted.as_mut(), empty_expr);
-                self.replace_unquote(expr, unquote, unquoted_exprs);
-            }
-        }
-    }
-
-    fn find_unquoted_exprs_in_literal(
-        &mut self,
-        literal: &mut Literal,
-        unquoted_exprs: &mut Vec<ExprId>,
-    ) {
-        match literal {
-            Literal::Array(array) | Literal::Slice(array) => match array {
-                ArrayLiteral::Standard(elements) => {
-                    for element in elements {
-                        self.find_unquoted_exprs_in_expr(element, unquoted_exprs);
+    /// Go through the given tokens looking for a '$' token followed by a variable to unquote.
+    /// Each time these two tokens are found, they are replaced by a new UnquoteMarker token
+    /// containing the ExprId of the resolved variable to unquote.
+    pub fn find_unquoted_exprs_tokens(&mut self, tokens: Tokens) -> Tokens {
+        let token_count = tokens.0.len();
+        let mut new_tokens = Vec::with_capacity(token_count);
+        let mut tokens = tokens.0.into_iter();
+
+        while let Some(token) = tokens.next() {
```

Comment:
This is with iterators, but I don't think there's any clean way to do it in a for loop. Since we need mutable access to the iterator to only conditionally grab two elements if a `$` is seen in the stream.